### PR TITLE
Add ability to declare QuestLists in .dfmod file.

### PR DESCRIPTION
**Pull request by Jagget**
Currently you can activate new Quest Pack (quest list) by adding the file to StreamingAssets\QuestPacks\ folder.

In case of packing quests in .dfmod file, you must call `QuestListsManager.RegisterQuestList("MyQuests")`.

It is not possible to ship quests in .dfmod without C# script.

I propose to register them in "Contributes" ModInfo section, the same way as BooksMappings 

Example:

```json

{
    "ModTitle": "QuestPack",
    "ModVersion": "1.0.0",
    "Files": [
        "QUEST1.txt",
        "QUEST2.txt",
        "QUEST3.txt",
        "QuestList-MyQuests.txt"
    ],
    "Contributes": {
        "QuestLists": [
            "MyQuests"
        ]
    }
}

```
image

image

Screenshot 2024-02-10 212729

### CODE REVIEW

**Comment by ajrb** [comment]
My thinking behind this was that if you were packaging quests into a mod then you'd very likely have some c# script even if just to set things up and act on settings etc. The quest pack idea was because Jay (and others) didn't want to create a dfmod package and script files etc for purely quests. Given some of the abuse of DFU quest packs have enabled I am not sure it was a good idea, but there you go. :-)

This may have been a better way to go than questpacks, but unless we intend to deprecate and remove quest pack functionality I don't really see what this achieves. What am I missing?

**Comment by KABoissonneault** [comment]
> My thinking behind this was that if you were packaging quests into a mod then you'd very likely have some c# script even if just to set things up and act on settings etc. The quest pack idea was because Jay (and others) didn't want to create a dfmod package and script files etc for purely quests. 

The culture has evolved, even quest-only modders can press Build on the Mod Builder nowadays, without any scripting knowledge. 

> Given some of the abuse of DFU quest packs have enabled I am not sure it was a good idea, but there you go. :-)

I'm with you in avoiding mods using quests to create unrelated systems, but DFU has to support them regardless.

Outside of that, my only issue with the PR is that it seems to require modifying the JSON manually to enable the feature, which is a step better than C#, but should probably be a Mod Builder feature. Granted, I think we can push that part to a later PR

**Comment by Jagget** [comment]
> Outside of that, my only issue with the PR is that it seems to require modifying the JSON manually to enable the feature, which is a step better than C#, but should probably be a Mod Builder feature. Granted, I think we can push that part to a later PR

It is actually the same for Bookmappings and Spellicons. There's no editor for "Contributes" part of the `dfmod.json`

**Comment by KABoissonneault** [comment]
Yeah I recognize that it's an existing issue, we'll consider it later if users want one.

I want to add that in terms of having QuestPacks in .dfmod, I have observed users confused as to why Quest Pack 1 and other quest mods were not showing up in the DFU mod list. I think moving forward, I would prefer if all mods came in .dfmod files, and keep loose files for development and very specific user configuration purposes. So this PR has my approval

File: Assets/Game/Addons/ModSupport/ModManager.cs
Comment (KABoissonneault) [inline Assets/Game/Addons/ModSupport/ModManager.cs:961]: For strings, it makes more sense to check `string.IsNullOrEmpty(directory)`. Especially since Path.GetDirectoryName is specified to return an empty string sometimes: https://learn.microsoft.com/en-us/dotnet/api/system.io.path.getdirectoryname

File: Assets/Game/Addons/ModSupport/ModManager.cs
Comment (KABoissonneault) [inline Assets/Game/Addons/ModSupport/ModManager.cs:951]: Could have used a defaulted =false, but worst case it breaks someone's Editor tools, it doesn't break any mods

Comment (Jagget) [reply Assets/Game/Addons/ModSupport/ModManager.cs]: Fixed

**Comment by ajrb** [comment]
Thanks for the answers.  Looks good to me.

**Review by ajrb — DISMISSED** [review]

File: Assets/Game/Addons/ModSupport/ModManager.cs
Comment (KABoissonneault) [inline Assets/Game/Addons/ModSupport/ModManager.cs:956]: I think this function should make `modInfo.Contributes.QuestLists` empty if `automaticallyRegisterQuestLists` is false. Right now, you can't disable the feature from the UI. Unchecking the box and resaving does nothing, it's gonna be checked again next time you open it, and the quests will be registered, because they are in the contributes still

**Comment by KABoissonneault** [comment]
Having finally tested this, the feature is almost all good, save one last outstanding issue. Gonna approve after this

**Review by KABoissonneault — APPROVED** [review]

**Review by ajrb — APPROVED** [review]

**Pull request merged**
